src/engagement_db_to_analysis/code_imputation_functions.py

```diff
@@ -134,26 +150,134 @@ def _impute_age_category(user, messages_traced_data, analysis_dataset_configs):
+    else:
+        return scheme.get_code_with_match_value(clean_value)
+
+
+def _impute_kenya_location_codes(user, messages_traced_data, analysis_dataset_configs):
+    """
+    Imputes Kenya location labels for location dataset messages.
+
+    :param user: Identifier of user running the pipeline.
+    :type user: str
+    :param messages_traced_data: Messages TracedData objects to impute age_category.
+    :type messages_traced_data: list of TracedData
+    :param analysis_dataset_configs: Analysis dataset configuration in pipeline configuration module.
+    :type analysis_dataset_configs: pipeline_config.analysis_configs.dataset_configurations
+    """
+
+    log.info(f"Imputing Kenya location labels for location messages...")
+
+    # Get the coding configurations for constituency and county analysis datasets
+    constituency_coding_config = None
+    county_coding_config = None
+    location_engagement_db_datasets = None
+    for analysis_dataset_config in analysis_dataset_configs:
+        for coding_config in analysis_dataset_config.coding_configs:
+            if coding_config.kenya_analysis_location is None:
+                log.info(f"No kenya_location config in {coding_config.analysis_dataset} skipping...")
+
+            elif coding_config.kenya_analysis_location == KenyaAnalysisLocations.CONSTITUENCY:
+                log.info(f"Found kenya_analysis_location in {coding_config.analysis_dataset} coding config")
+
+                assert constituency_coding_config is None, f"Found more than one constituency_coding_config in " \
+                    f"analysis_dataset_config"
+                constituency_coding_config = coding_config
+                location_engagement_db_datasets = analysis_dataset_config.engagement_db_datasets
+
+            elif coding_config.kenya_analysis_location == KenyaAnalysisLocations.COUNTY:
+                log.info(f"Found kenya_analysis_location in {coding_config.analysis_dataset} coding config")
+
+                assert county_coding_config is None, f"Found more than one county_coding_config in " \
+                    f"analysis_dataset_config"
+                county_coding_config = coding_config
+
+    assert county_coding_config is not None, f"No county_coding_config in " \
+        f"analysis_dataset_config"
+    assert constituency_coding_config is not None, f"No constituency_coding_config in " \
+        f"analysis_dataset_config"
+
+    for message_traced_data in messages_traced_data:
+        if message_traced_data["dataset"] in location_engagement_db_datasets:
+            message = Message.from_dict(dict(message_traced_data))
+            message_analysis_config = analysis_dataset_config_for_message(analysis_dataset_configs, message)
+
+            # Up to 1 location code should have been assigned in Coda. Search for that code,
+            # ensuring that only 1 has been assigned or, if multiple have been assigned, that they are non-conflicting
+            # control codes
```

Comment:
Can you note that multiple normal codes will also be converted to Coding Error, even if they were compatible (e.g. langata + nairobi)

Reply:
updated thanks.